common/lib/xmodule/xmodule/docs/decisions/0001-move-due-dates-to-db.rst

```diff
@@ -0,0 +1,52 @@
+
+Status
+------
+
+Proposed
+
+Context
+-------
+
+We want an authoritative and readily-accessible place to track due dates for sections/units in a course.  We also want to override due dates per learner, and record the reason for the override. The current implementation stores due dates in the XBlock (in mongodb). If you want to know all of the due dates in a course, you have to walk the entire course structure and pull out each date, which is an inefficient operation.
+
+There is currently a way to override due dates per learner, using `IDDE <https://github.com/mitodl/ccx-idde-overrides-slides/blob/master/markdown/slides.md#individual-due-date-extensions-idde>`_, but this relies on a generic field-override system which doesn't know anything about the underlying data, and enabling this feature results in a database query for every xblock access, for every course on the platform. IDDE also does not allow for an audit trail on due date extensions, nor does it allow instructors to record a reason for the extension. 
+
+
+Decisions
+---------
+
+1. Create a new pluggable Django app responsible for dates and date overrides
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This approach anticipates other date-related functions in the future, such as APIs for returning a calendar view of upcoming dates.
+
+- The app will contain at least these models:
+    + ``ContentDate``: mapping content ids to dates
+    + ``UserContentDate``: mapping user ids to ``ContentDates``. It will also record an audit trail for the override.
+- The app will contain a custom XBlock field that will be a drop-in replacement for the existing Date XBlock field.
+    + If the date exists in the relational database (whether in UserContentDate or ContentDate), it'll use that date; otherwise, it'll use the one stored in the XBlock.
+- The app will expose a REST API for retrieving and updating dates in the database.
+
+2. Create an instructor interface for setting student due dates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The existing IDDE instructor interface may be used, or a modern micro-frontend can be included from the new pluggable Django app.
+
+3. All dates for a course/learner will be retrieved using one query, cached for the current request.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It's important that the underlying implementation can efficiently retrieve all due dates for the course, otherwise it's not an improvement over IDDE.
+
+4. All new courses will automatically start storing due dates in the database.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Since the new XBlock field will transparently convert dates between the content store and relational database, importing and exporting courses should automatically work. If not, an explicit import/export step may be needed. 
```

Comment:
What if we leave date storage exactly as it is today in Studio, but have an updated model that we build during the publish phase for the LMS? We can then do a little ugly magic under the covers so that field data for the LMS XBlock runtime returns the right thing for the relevant fields (definitely `due`, but we might also want `start`).

This might make our lives easier in a few ways:

* We need that translation layer for compatibility anyway, and this way we can have only one way to write to the ContentDate model (well, for now).
* Since we derive a new data source from Studio input, we can leave the current import/export flow alone, reducing the odds of breaking compatibility, and letting us avoid that whole two way data sync issue.
* All the real performance concerns are on the LMS side anyway.
* All per-user overrides happen in the LMS anyway.
* Because we're building a parallel system, it's much easier to toggle on/off, or to dry run test it for a while to make sure it's returning the same results.
* It gives a pattern for gradually moving over more bits of scheduling over time, without a big up-front commitment.

Reply:
I like @ormsbee's suggestion here.  It should be relatively lightweight and directionally aligned with encapsulating course-administration behavior within the LMS and outside of Studio.

Reply:
+1, I like this idea. I'm not a big fan of the "virtual due date XBlock field" approach since that seems like a special case.

I also think that we really need to move away from storing absolute dates with the content (in modulestore/xblock/blockstore). Currently, if you re-run a course, [you need to manually change all the start & due dates to the new \[semester\]'s schedule](https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/named-release-cypress/building_course/course_rerun.html#data-duplicated-when-you-re-run-a-course). Likewise for CCX, if you e.g. create a course that runs two weeks behind the master course, you have to manually adjust each date by two weeks.

So I think in the future, the content should store a `relative_due` field which stores the date relative to the course start date, in lieu of the `due` field. Then, the author can define things like "Section 1 Assignment 4 is due two weeks after Assignment 3" etc., and re-runs and CCX courses will automatically have the recommended due date schedule when created.

With the approach @ormsbee is suggesting here, I think that might be a relatively easy incremental change, because one XBlock at a time could change from `due` to `relative_due`, and the publish->ContentDate step could use  `relative_due` if available, or fall back to `due`.